### PR TITLE
fix DynamicPartitionKafkaInputDStream checkpint and add C# func support

### PR DESCRIPTION
**Pull request by xiongrenyi**
1. fix DynamicPartitionKafkaInputDStream checkpoint by fixing restore method and starting streaming callback server ealier during restore time
2. add C# func support in DynamicPartitionKafkaInputDStream by inheriting from InputDStream directly

### CODE REVIEW

File: scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala [hunk 1]
```diff
@@ -102,7 +102,7 @@ class CSharpDStream(
   override def slideDuration: Duration = parent.slideDuration
 
   override def compute(validTime: Time): Option[RDD[Array[Byte]]] = {
-    val rdd = parent.compute(validTime)
+    val rdd = parent.getOrCompute(validTime)
```
Comment (skaarthik) [inline scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala]: What is the reason for this change?

File: scala/src/test/scala/org/apache/spark/streaming/api/kafka/DynamicPartitionKafkaInputDStreamSuite.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.spark.streaming.kafka
+
+import kafka.common.TopicAndPartition
+import kafka.message.MessageAndMetadata
+import kafka.serializer.DefaultDecoder
+import org.apache.spark.csharp.SparkCLRFunSuite
+import org.apache.spark.streaming.{Duration, StreamingContext, Time}
+import org.apache.spark.{Partition, SparkConf, SparkContext}
+
+class DynamicPartitionKafkaInputDStreamSuite extends SparkCLRFunSuite {
```
Comment (skaarthik) [inline scala/src/test/scala/org/apache/spark/streaming/api/kafka/DynamicPartitionKafkaInputDStreamSuite.scala]: Good to see a new test suite. Can you also add test for compute method?

File: scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala (hunk 1 above)
Comment (xiongrenyi) [inline scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala]: it's a bug I think.
getOrCompute put generated RDDs in a HashMap for future reference including checkpoint.
double checked python did same thing.
not sure why it was compute.

Comment (skaarthik) [inline scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala]: Ok, missing test is probably one of the reasons for this bug left uncovered. Can we add a test now?

**Comment by xiongrenyi** [comment]
Added by calling CSharpDStream.compute

File: scala/src/test/scala/org/apache/spark/streaming/api/kafka/DynamicPartitionKafkaInputDStreamSuite.scala (hunk 2 above)
Comment (xiongrenyi) [inline scala/src/test/scala/org/apache/spark/streaming/api/kafka/DynamicPartitionKafkaInputDStreamSuite.scala]: Added

**Comment by qintao1976** [comment]
LGTM
